### PR TITLE
nixos/gitlab: Fix state directory permissions, clean up initializers directory

### LINKED ISSUE #68696: gitlab module permission denied after upgrade
**Describe the bug**
#66274 breaks existing GitLab installations.
In a 19.03 GitLab installation, in the state directory…
  * `/gitlab_shell_secret` is owned by root and writable by only root
  * `/config/secrets.yml` and `/config/database.yml` are linked to the nix store

The new startup script from #66274 attempts to write all of these files, but has insufficient permissions because it is running as `gitlab` and the two ymls are read only.

In order to upgrade, you have to:
* `chown gitlab:gitlab /gitlab_shell_secret`
* `rm /config/database.yml`
* `rm /config/secrets.yml`

This change should be backwards compatible or the migration path should be documented.

cc @talyz @flokli 

Great work by the way, my store does not have any secrets anymore :)

### PR DESCRIPTION
**Pull request by talyz**
###### Motivation for this change
Since the preStart script is no longer running in privileged mode, recursively reassign all files in the state directory to the user we're running as. Also, delete the database.yml symlink if it exists,
since we want to create a real file in its place.

I've tested this by manually changing the owner of `gitlab_shell_secret`, and symlinking `database.yml` and `secrets.yml` to an irrelevant file. Additional testing would be appreciated.

The `config/initializers` directory is populated with files from the gitlab distribution on start, but old files will be left in the state folder even if they're removed from the distribution, which can lead to
startup failures. Fix this by always purging the directory on start before populating it.

Should hopefully fix #68696; if so, it needs to be pulled into 19.09 too.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @flokli @globin @aanderse @FRidh

### CODE REVIEW

File: nixos/modules/services/misc/gitlab.nix [hunk 1]
```diff
@@ -774,6 +775,10 @@ in {
 
           ${pkgs.openssl}/bin/openssl rand -hex 32 > ${cfg.statePath}/gitlab_shell_secret
 
+          if [[ -h '${cfg.statePath}/config/database.yml' ]]; then
+            rm '${cfg.statePath}/config/database.yml'
+          fi
+
           ${if cfg.databasePasswordFile != null then ''
```
Comment (schneefux) [inline nixos/modules/services/misc/gitlab.nix]: Not related to this issue, but I noticed that no `database.yml` gets written if the `databasePasswordFile` is not set. So it is impossible to exclusively use `extraDatabaseConfig` to configure the database.
In other modules, it is sometimes easier to ditch the options and put a complete config in the `extraConfig`.
Is that intentional?

File: nixos/modules/services/misc/gitlab.nix [hunk 2]
```diff
@@ -774,6 +775,10 @@ in {
 
           ${pkgs.openssl}/bin/openssl rand -hex 32 > ${cfg.statePath}/gitlab_shell_secret
```
Comment (schneefux) [inline nixos/modules/services/misc/gitlab.nix]: I chowned `gitlab_shell_secret` back to `root:root` and it breaks here.
` 4yjg33mxs50ngydn1glpb9zcx15i7np1-unit-script-gitlab-pre-start[24046]: /nix/store/4yjg33mxs50ngydn1glpb9zcx15i7np1-unit-script-gitlab-pre-start: line 14: /var/www/gitlab/gitlab_shell_secret: Permission denied`

**Review by schneefux — CHANGES_REQUESTED** [review]
I have tested that this does not break an upgraded installation. If I link the two ymls to random files, they are replaced correctly. 👍 
However, if the shell secret is owned by a different account than `gitlab`, the start script still fails.

File: nixos/modules/services/misc/gitlab.nix (hunk 2 above)
Comment (talyz) [reply nixos/modules/services/misc/gitlab.nix]: Yes, this hinges on systemd-tmpfiles getting to do its job before the script starts, which it should on reboot or when the configuration is activated, if I'm not mistaken. You can manually try it with `systemd-tmpfiles --create`.

File: nixos/modules/services/misc/gitlab.nix (hunk 1 above)
Comment (talyz) [reply nixos/modules/services/misc/gitlab.nix]: If that's the case, then that's a bug, but I don't have `databasePasswordFile` set and get a `database.yml` anyway (see the else clause of the if expression)

**Review by schneefux — APPROVED** [review]
👍

**Comment by globin** [comment]
Please don't do this as this takes ages on larger gitlab installations (~15 mins for us)..
I'd prefer fixing this for now with chown and running `preStart` as root again for 19.09 as a compatibility release with the right permissions and switching to `preStart` as gitlab in 20.03

File: nixos/modules/services/misc/gitlab.nix (hunk 1 above)
Comment (talyz) [reply nixos/modules/services/misc/gitlab.nix]: No problem :)

**Comment by aanderse** [comment]
If you restored the `tmpfiles` rules from prior to these changes but set `tmpfiles` mode to `z` for the individual files owned by `root` that would save a recursive `chown` + `chmod`, allow you to run the script unprivileged, and not require any further changes at a later time, I believe.

**Comment by talyz** [comment]
@globin Oh, for every config activation or just the first one? I guess it makes sense, since it has to stat all files recursively... :/

@aanderse That was my first thought, but it didn't work - systemd-tmpfiles emits the error message
> Detected unsafe path transition /var/gitlab/state → /var/gitlab/state/gitlab_shell_secret during canonicalization of /var/gitlab/state/gitlab_shell_secret.

which it apparently does when a file in a directory owned by an unprivileged user is owned by root...

**Comment by talyz** [comment]
Okay, so I've gone with the earlier approach of running a privileged part and an unprivileged part of the `preStart` script, but implemented it locally in the gitlab module rather than the systemd module. Does this work better for you, @globin?

**Comment by talyz** [comment]
I've pushed one additional commit that cleans up the `config/initializers` directory, rebased on master and tested it without any issues. This addresses the concern raised by @bgamari in https://github.com/NixOS/nixpkgs/pull/70216#issuecomment-537646392.

@globin Do you have any issues with the current approach or could this be merged?

**Review by aanderse — APPROVED** [review]
I have not tested, nor ever used the `gitlab` module on NixOS but I understand and approve of the basic idea behind this change. From my limited perspective the actual implementation looks fine as well. :+1:

**Pull request merged**